### PR TITLE
[contrib + e2e] prometheus: add pushgateway functionality to prometheus allinone deployment along w/ runRC Gauge reporting.

### PR DESCRIPTION
**Pull request by jayunit100**
- Add prom push gateway
- Update promehteus args to scan ^ 
- Update `util.go` to write running/pending metrics into the gateway during tests if provided. 
  This now allows us to push to the 9091 gateway from any tests, and prometheus will automatically collect them over the service endpoint.

### CODE REVIEW

**Comment by timothysc** [comment]
Contrib, so it should meet the criteria for entry.

**Comment by jayunit100** [comment]
screenshot from 2015-07-09 18-23-03

**Comment by jayunit100** [comment]
okay, ready for review now :) !

**Comment by jayunit100** [comment]
tested that this works with the exact pod specification changes in this PR, just do `kubectl create -f pushgateway-service.json ; cluster/kubectl.sh create -f ./prometheus-service.json ; kubectl create -f ./prometheus-all.json`  and then wait a while, before running e2e's like so `_output/local/bin/linux/amd64/e2e.test --provider=local --host=http://127.0.0.1:8080 -ginkgo.focus=Density --kubeconfig=/opt/kube-creds --prom-push-gateway="http://127.0.0.1:9091"`

File: contrib/prometheus/prometheus-all.json [hunk 1]
```diff
@@ -62,6 +59,23 @@
...
+                            {
+                                "name": "PUSHGATEWAY_SERVICE_HOST",
+                                "value": "localhost"
+                            },
+                            {
+                                "name": "PUSHGATEWAY_SERVICE_PORT",
+                                "value": "9091"
+                            }
+                        ],
+                        "image": "jayunit100/kube-prometheus",
```
Comment (timothysc) [inline contrib/prometheus/prometheus-all.json]: side note: Should we talk with prom upstream to mock their docker files for kube?

File: test/e2e/util.go [hunk 2]
```diff
@@ -993,6 +995,22 @@ func Diff(oldPods []*api.Pod, curPods []*api.Pod) PodDiff {
 // namespace lifecycle for handling cleanup).
 func RunRC(config RCConfig) error {
 
+	// Prometheus stuff: Setup metrics.
+	runningMetric := prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (timothysc) [inline test/e2e/util.go]: don't you want to if{wrap} this if the arg exists or not?

File: test/e2e/util.go [hunk 3]
```diff
@@ -1102,6 +1122,20 @@ func RunRC(config RCConfig) error {
 
 		Logf("%v %v Pods: %d out of %d created, %d running, %d pending, %d waiting, %d inactive, %d unknown ",
 			time.Now(), rc.Name, len(pods), config.Replicas, running, pending, waiting, inactive, unknown)
+
+		runningMetric.Set(float64(running))
```
Comment (timothysc) [inline test/e2e/util.go]: I would probably push this into the if block.

**Comment by timothysc** [comment]
Eventually if things go well I could see an abstraction taking place to wrap any prometheus interface calls, as folks could push to something else as well. 

But for now it's probably fine and we can iterate as needed.

File: test/e2e/util.go (hunk 2 above)
Comment (jayunit100) [inline test/e2e/util.go]: the push is orthagonal to the gauge, so i could if you want me to, but not sure its needed

**Comment by jayunit100** [comment]
Agree on the abstraction.  it feels a little one-offish.  and i think the splitting of the code in the method might confuse folks.

**Comment by timothysc** [comment]
@fgrzadkowski, @wojtek-t this may be of interest to you.

File: contrib/prometheus/prometheus-all.json (hunk 1 above)
Comment (jayunit100) [inline contrib/prometheus/prometheus-all.json]: yeah and this one needs to be updated, am talking to them

File: test/e2e/util.go (hunk 3 above)
Comment (jayunit100) [inline test/e2e/util.go]: @timothysc okay to solve this later when we abstract metrics further?  For now gaurding in the if statement probably won't buy us anything...

**Comment by jayunit100** [comment]
Hi folks.  Anyone want to look at this guy ? We definetly want to run it on our E2Es daily, maybe @nikhiljindal  @lavalamp

**Comment by lavalamp** [comment]
Can this wait til post v1? We're still trying to reduce churn even in our test code for at least another week.

File: test/e2e/util.go [hunk 4]
```diff
@@ -1137,6 +1157,20 @@ func RunRC(config RCConfig) error {
...
+		pendingMetric.Set(float64(pending))
+
+		if testContext.PrometheusPushGateway != "" {
```
Comment (lavalamp) [inline test/e2e/util.go]: Is there a way to make this work without sprinkling these if blocks all over our code base? This is going to be hard to read; also we definitely will need some tests if you want this to keep working.

Comment (jayunit100) [inline test/e2e/util.go]: I'll look into a more modular implementation and feel free to punt to post 1.0 I understand the churn issue

**Comment by jayunit100** [comment]
UNTESTED DO NOT COMMIT.  Will test later on.

**Comment by mikedanese** [comment]
Not sure who to assign this to. Feel free to rebalance.

**Comment by timothysc** [comment]
@mikedanese , @fgrzadkowski is probably the best candidate, he's dealt with both topics.

**Comment by mikedanese** [comment]
I think he is OOO for a week or more. @wojtek-t feel free to assign to him if i'm wrong or you can't get to it in that timeframe

**Comment by wojtek-t** [comment]
Will look into it today

File: contrib/prometheus/README.md [hunk 5]
```diff
@@ -29,18 +29,25 @@ Looks open enough :).
...
-Then (edit the publicIP field in prometheus-service to be a public ip on one of your kubelets), 
+This will launch the pushgateway such that prometheus is collecting metrics from it automatically.
 
-and run 
+NOTE: If you plan to use the push gateway to push metrics (for example, we use it in e2e tests) 
 
-`kubectl create -f contrib/prometheus/prometheus-service.json`
+you will want to have a stable IP that you can provide the push gateway from, and you may need to modify the NodePort stuff
+
+for your cloud provider.
```
Comment (wojtek-t) [inline contrib/prometheus/README.md]: please move it to previous line

File: contrib/prometheus/README.md [hunk 6]
```diff
@@ -69,11 +76,12 @@ By default the metrics are written to a temporary location (that can be changed
 in the the volumes section of the yaml file). Prometheus' UI is available 
 at port 9090.
 
+
+
```
Comment (wojtek-t) [inline contrib/prometheus/README.md]: I don't think we need 3 empty lines here

File: contrib/prometheus/prometheus-all.json [hunk 7]
```diff
@@ -21,14 +21,14 @@
...
-                        "name":"kube-promdash",
-                        "image": "prom/promdash",
                         "env": [
                             {
                                 "name": "DATABASE_URL",
                                 "value": "sqlite3:/promdash/file.sqlite3"
                             }
                         ],
+                        "image": "prom/promdash",
+                        "name": "kube-promdash",
```
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: Why did you move that - I think having name and image at the beginning is more intuitive.

File: contrib/prometheus/prometheus-all.json (hunk 1 above)
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: did you resolve that?

File: contrib/prometheus/prometheus-all.json [hunk 8]
```diff
@@ -72,11 +86,25 @@
                         ]
                     },
                     {
-                        "name": "kubectl",
-                        "image": "gcr.io/google_containers/kubectl:v0.18.0-350-gfb3305edcf6c1a",
-                        "args": [
-                            "proxy", "-p", "8001", "--api-prefix=/"
+                        "image": "jayunit100/prompush:0.2.0",
```
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: is is possible to make the official image for this and store it in gcr?

File: contrib/prometheus/prometheus-all.json [hunk 9]
```diff
@@ -72,11 +86,25 @@
...
+                    },
+                    {
+                        "args": [
+                            "proxy",
+                            "-p",
+                            "8001",
+                            "--api-prefix=/"
+                        ],
+                        "image": "gcr.io/google_containers/kubectl:v0.18.0-350-gfb3305edcf6c1a",
+                        "name": "kubectl"
```
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: nit: can you move image and name to the top (i.e. before args)?

File: test/e2e/prompush.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (wojtek-t) [inline test/e2e/prompush.go]: s/2014/2015/

File: test/e2e/prompush.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	Name: "e2e_running",
+	Help: "The num of running pods",
+})
+
+var pendingMetric = prometheus.NewGauge(prometheus.GaugeOpts{
+	Name: "e2e_pending",
+	Help: "The num of pending pods",
+})
+
+//Turn this to true after we register.
```
Comment (wojtek-t) [inline test/e2e/prompush.go]: nit: s/Turn/ Turn/ (space before Turn)

File: test/e2e/prompush.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+var prom_registered = false
+
+// Reusable function for pushing metrics to prometheus.  Handles initialization and so on.
+func promPushRunningPending(running, pending int) error {
+
+	if testContext.PrometheusPushGateway == "" {
+		Logf("Ignoring prom push, push gateway unavailable")
+		return nil
+	} else {
+		//Register metrics if necessary
```
Comment (wojtek-t) [inline test/e2e/prompush.go]: same here and below

File: contrib/prometheus/prometheus-all.json (hunk 8 above)
Comment (jayunit100) [inline contrib/prometheus/prometheus-all.json]: happy to do so.  Can we do in a follow on PR ? The reason being that there are other images i'd like to add to Gcr and i don't know who will / how to  push them, so its a independent effort.

Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: Sure - but please add a TODO for it.

File: contrib/prometheus/prometheus-all.json (hunk 1 above)
Comment (jayunit100) [inline contrib/prometheus/prometheus-all.json]: resolved below ~ we'll put it in GCR in follow in issue, will create that issue now...

File: contrib/prometheus/prometheus-all.json (hunk 9 above)
Comment (jayunit100) [inline contrib/prometheus/prometheus-all.json]: maybe we can add this to verify-fmt ? its a good convention that your suggestion.  ill fix it now

**Comment by jayunit100** [comment]
fixed ! @wojtek-t .  thanks for the review.  i created #11808 for the follow on GRC container push.

File: contrib/prometheus/prometheus-all.json [hunk 13]
```diff
@@ -21,15 +21,15 @@
...
                         "image": "prom/promdash",
+                        "name": "kube-promdash",
                         "env": [
                             {
                                 "name": "DATABASE_URL",
                                 "value": "sqlite3:/promdash/file.sqlite3"
                             }
                         ],
-                        "ports": [
+                                             "ports": [
```
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: fix indentation please

File: contrib/prometheus/prometheus-all.json (hunk 1 above)
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: please move "image" and "name" to the top (as in other places)

**Comment by wojtek-t** [comment]
just two more small comments

**Comment by jayunit100** [comment]
okidokie, fixed and squashed

**Comment by mikedanese** [comment]
@k8s-bot ok to test. add to whitelist

**Comment by jayunit100** [comment]
Services test failed somehow.  Could that have been cause by this pr? Or is that just a flake.

**Comment by wojtek-t** [comment]
>
Services test failed somehow. Could that have been cause by this pr? Or is that just a flake.

I'm almost sure this is unrelated flake.

Rerunning Jenkins tests.

**Comment by wojtek-t** [comment]
@k8s-bot test this please

File: contrib/prometheus/prometheus-all.json [hunk 14]
```diff
@@ -62,6 +59,23 @@
...
+                                "name": "PUSHGATEWAY_SERVICE_HOST",
+                                "value": "localhost"
+                            },
+                            {
+                                "name": "PUSHGATEWAY_SERVICE_PORT",
+                                "value": "9091"
+                            }
+                        ],
+                        "image": "jayunit100/kube-prometheus",
+                        "name": "kube-prometheus",
```
Comment (wojtek-t) [inline contrib/prometheus/prometheus-all.json]: please move "image" and "name" to the top (as in other places)

**Comment by jayunit100** [comment]
ok . just did the minor reformat on the json......

**Comment by wojtek-t** [comment]
LGTM

**Comment by timothysc** [comment]
w00t

**Comment by jayunit100** [comment]
Since this is LGTM, Is it going to merge today?

**Comment by ahmetb** [comment]
We moved the examples to their own repo (https://github.com/kubernetes/examples) for further maintenance. It looks like this issue has been open with little to no recent activity, if you think it's still relevant, please create a new issue in the new repo.

/assign
